midend/predication.cpp

```diff
@@ -38,41 +39,37 @@ static cstring lvalue_name(const IR::Expression *exp) {
     return cstring();
 }
 
-bool Predication::EmptyStatementRemover::preorder(IR::BlockStatement * block) {
-    for (auto iter = block->components.begin(); iter != block->components.end();) {
-        if ((*iter)->is<IR::EmptyStatement>()) {
-            iter = block->components.erase(iter);
-        } else if ((*iter)->is<IR::BlockStatement>()) {
-            visit(*iter);
-            if ((*iter)->to<IR::BlockStatement>()->components.size() == 0) {
-                iter = block->components.erase(iter);
-            } else {
-                iter++;
-            }
-        } else {
-            iter++;
-        }
-    }
-    return false;
+const IR::Node* Predication::EmptyStatementRemover::postorder(IR::EmptyStatement* statement) {
+    return nullptr;
 }
 
+const IR::Node* Predication::EmptyStatementRemover::postorder(IR::BlockStatement* statement) {
+    if (statement->components.empty()) return nullptr;
```

Comment:
Note that sometimes empty blocks and empty statements cannot be removed safely, for example, if they are the only statement of an 'if-then' branch. Hopefully you know that this case cannot occur.

Reply:
Yes I'm aware of this, I implemented this class based on a pass that does a similar simplification and figured out that most checks it had are not needed here since the only empty statements and blocks that are being removed here are the ones that the Predication code itself created, so no problematic cases can occur.